ops/model.py

```diff
@@ -3334,35 +3346,51 @@ def _run(
         use_json: bool = False,
         input_stream: Optional[str] = None,
     ) -> Union[str, Any, None]:
-        kwargs = {
-            'stdout': subprocess.PIPE,
-            'stderr': subprocess.PIPE,
-            'check': True,
-            'encoding': 'utf-8',
-        }
-        if input_stream:
-            kwargs.update({'input': input_stream})
-        which_cmd = shutil.which(args[0])
-        if which_cmd is None:
-            raise RuntimeError(f'command not found: {args[0]}')
-        args = (which_cmd,) + args[1:]
-        if use_json:
-            args += ('--format=json',)
-        # TODO(benhoyt): all the "type: ignore"s below kinda suck, but I've
-        #                been fighting with Pyright for half an hour now...
-        try:
-            result = subprocess.run(args, **kwargs)  # type: ignore
-        except subprocess.CalledProcessError as e:
-            raise ModelError(e.stderr) from e
-        if return_output:
-            if result.stdout is None:  # type: ignore
-                return ''
-            else:
-                text: str = result.stdout  # type: ignore
-                if use_json:
-                    return json.loads(text)  # type: ignore
+        if self._is_recursive.get():
+            # Would be nice to do something about this, sys.stderr maybe?
```

Comment:
If this happens, it would be considered an ops bug, right? If so, then I think a warning would be good - of course we shove those into the logging module too, so you'd have to do that carefully.

Reply:
Either ops bug or `juju-log` fails.

I occasionally have juju hook tools fail when the laptop is suspended and resumed. I'm not sure if that translates to production in any way though.

Reply:
reworded